### PR TITLE
libsubprocess: fix bugs and clean up subprocess server

### PR DESCRIPTION
**Pull request by garlick**
Problem: the subprocess server used in the broker does not properly terminate streaming RPCs, does not protect rank 0 against remote kill request, and contains idioms from an earlier era of flux development that can now be reworked and simplified.

This PR contains a lot of cleanup, but the more impactful changes are:
- the server protocol is simplified, and now conforms to RFC 6
- rank 0 authorization issues are fixed
- the server API is no longer public, and it looks more like modern flux interfaces
- server shutdown is now performed before the reactor exits

### CODE REVIEW

File: src/broker/exec.c [hunk 1]
```diff
@@ -30,34 +30,34 @@
...
     /* exec_initialize() never called */
     if (!s)
         return;
 
-    if (flux_subprocess_server_subprocesses_kill (s,
+    if (subprocess_server_subprocesses_kill (s,
                                                   SIGTERM,
                                                   EXEC_TERMINATE_TIMEOUT) < 0)
-        flux_log_error (h, "flux_subprocess_server_subprocesses_kill");
+        flux_log_error (h, "subprocess_server_subprocesses_kill");
```
Comment (chu11) [inline src/broker/exec.c]: nit: in ` libsubprocess: drop flux_ prefix from server API ` correct tabbing (but appears code will disappear in later commit)

File: src/broker/exec.c [hunk 2]
```diff
@@ -30,34 +30,34 @@
...
      * messages while we begin teardown, so we will SIGKILL the
      * subprocesses as well.
      */
 
-    if (flux_subprocess_server_subprocesses_kill (s,
+    if (subprocess_server_subprocesses_kill (s,
                                                   SIGKILL,
                                                   EXEC_TERMINATE_TIMEOUT) < 0)
-        flux_log_error (h, "flux_subprocess_server_subprocesses_kill");
+        flux_log_error (h, "subprocess_server_subprocesses_kill");
```
Comment (chu11) [inline src/broker/exec.c]: nit: in `libsubprocess: drop flux_ prefix from server API` correct tabbing (but appears code will disappear in later commit)

File: src/common/libsubprocess/server.c [hunk 3]
```diff
@@ -629,10 +629,10 @@ static void server_disconnect_cb (flux_t *h,
     const char *sender;
 
     if ((sender = flux_msg_route_first (msg)))
-        server_terminate_by_uuid (s, sender);
+        subprocess_server_terminate_by_uuid (s, sender);
 }
```
Comment (chu11) [inline src/common/libsubprocess/server.c]: in ` libsubprocess/server: make functions local `, this change is in the wrong commit?

Comment (chu11) [reply src/common/libsubprocess/server.c]: oh i guess the function prototype is no longer defined b/c it's not in the header.  maybe commit message just needs tweaking, mentioning remove functions from header, adjust callers.

File: src/common/libsubprocess/server.c [hunk 4]
```diff
@@ -118,8 +118,10 @@ static void proc_completion_cb (flux_subprocess_t *p)
     proc_delete (s, p);
 }
 
-static void internal_fatal (subprocess_server_t *s, flux_subprocess_t *p)
+static void proc_internal_fatal (flux_subprocess_t *p)
```
Comment (chu11) [inline src/common/libsubprocess/server.c:118]: commit message typo "causal" -> "casual"

**Review by chu11 — COMMENTED** [review]
review through [libsubprocess/server: rename fatal error function](https://github.com/flux-framework/flux-core/pull/4944/commits/c0cbd49319af987f4a66829f4a2acb37cd2c2db9)

hopefully can look at rest tonight

File: src/common/libsubprocess/remote.c [hunk 5]
```diff
@@ -718,52 +718,6 @@ static void remote_exec_cb (flux_future_t *f, void *arg)
     p->f = NULL;
 }
 
-static void remote_continuation_cb (flux_future_t *f, void *arg)
```
Comment (chu11) [inline src/common/libsubprocess/remote.c:721]: commit message typo "anotehr" -> "another"

**Review by chu11 — COMMENTED** [review]
through [libsubprocess: respond with ENODATA not "complete"](https://github.com/flux-framework/flux-core/pull/4944/commits/62efb4b007d6572e5c4ebda59b0f20b9d4b3f0b3)

File: t/scripts/rexec.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+###############################################################
+# Copyright 2021 Lawrence Livermore National Security, LLC
```
Comment (chu11) [inline t/scripts/rexec.py:2]: 2023!

File: t/t0005-rexec.t [hunk 7]
```diff
@@ -241,4 +244,38 @@ test_expect_success 'rexec read_getline call works on remote streams' '
...
+
+test_expect_success NO_CHAIN_LINT 'ps, kill work locally on rank 0' '
+	$rexec sleep 3600 &
+	wait_rexec_process_count 1 0 &&
+	pid=$($rexec_script ps |cut -f1) &&
+	$rexec_script kill 15 $pid &&
+	wait_rexec_process_count 0 0
+'
+
+test_expect_success NO_CHAIN_LINT 'ps, kill is fail remotely on rank 0' '
```
Comment (chu11) [inline t/t0005-rexec.t:269]: "is fail" -> "fail"?

**Review by chu11 — APPROVED** [review]
oh wow, that's a lot of great cleanup.  As I was reading through the review I was like "why did I do that?" but you're right, there were a lot of things done differently when this was originally written.  And lots of minor fixes over time for minor things discovered, that did need a high level look and subsequent cleanup/refactor.

Approving, I think everything I found is just a nit or minor stuff.

File: src/broker/exec.c (hunk 1 above)
Comment (garlick) [reply src/broker/exec.c]: Eh, whoops.  I think this is probably not worth the effort to refactor the commits though, given that the PR doesn't leave behind a style issue.

**Comment by garlick** [comment]
Thanks @chu11!  I just pushed changes to address most of your comments, except I'd like to beg off doing the whitespace commit refactor since the problem you noticed is ultimately harmless (from the perspective of the PR as a whole).

**Comment by chu11** [comment]
oh yeah, should the title of the PR be prefixed with `libsubprocess`?

**Comment by garlick** [comment]
Setting MWP, thanks again.

**Pull request merged**